src/nnet3/discriminative-training.cc

```diff
@@ -0,0 +1,602 @@
+    index += supervision_.num_ali.size();
+  } 
+
+  KALDI_ASSERT(index == answers.size());
+  
+  if (nnet_output_deriv_) {
+    nnet_output_deriv_->SetZero();
+    KALDI_ASSERT(nnet_output_deriv_->NumRows() == nnet_output_.NumRows() &&
+        nnet_output_deriv_->NumCols() == nnet_output_.NumCols());
+  }
+
+  if (xent_output_deriv_) {
+    xent_output_deriv_->SetZero();
+    KALDI_ASSERT(xent_output_deriv_->NumRows() == nnet_output_.NumRows() &&
+        xent_output_deriv_->NumCols() == nnet_output_.NumCols());
+  }
+
+  Posterior post;
+  Posterior xent_post;
+  double objf = ComputeObjfAndDeriv(&post, 
+                (xent_output_deriv_ ? &xent_post : NULL));
+  
+  this_stats.tot_objf += supervision_.weight * objf;
+  
+  KALDI_ASSERT(nnet_output_.NumRows() == post.size());
+  
+  CuMatrix<BaseFloat> output_deriv;
+  
+  CuMatrixBase<BaseFloat> *output_deriv_temp; 
+  
+  if (nnet_output_deriv_) 
+    output_deriv_temp = nnet_output_deriv_;
+  else {
+    // This is for accumulating the statistics
+    output_deriv.Resize(nnet_output_.NumRows(), nnet_output_.NumCols());
+    output_deriv_temp = &output_deriv;
+  }
+  
+  double tot_num_post = 0.0, tot_den_post = 0.0;
+  {
+    std::vector<Int32Pair> deriv_indexes;
+    std::vector<BaseFloat> deriv_data;
+    for (size_t t = 0; t < post.size(); t++) {
+      for (size_t j = 0; j < post[t].size(); j++) {
+        int32 seq = t / supervision_.frames_per_sequence, 
+              idx = t % supervision_.frames_per_sequence;
+        int32 pdf_id = post[t][j].first;
+
+        // Same ordering as for 'chain' models
+        deriv_indexes.push_back(MakePair(idx * supervision_.num_sequences + seq, pdf_id));
+
+        BaseFloat weight = post[t][j].second;
+        if (weight > 0.0) tot_num_post += weight;
+        if (weight < 0.0) tot_den_post -= weight;
+        deriv_data.push_back(weight);
+      }
+    }
+    CuArray<Int32Pair> cu_deriv_indexes(deriv_indexes);
+    output_deriv_temp->AddElements(supervision_.weight, cu_deriv_indexes, deriv_data.data());
+  }
+
+  if (xent_output_deriv_) {
+    std::vector<Int32Pair> deriv_indexes;
+    std::vector<BaseFloat> deriv_data;
+    for (size_t t = 0; t < xent_post.size(); t++) {
+      for (size_t j = 0; j < xent_post[t].size(); j++) {
+        int32 seq = t / supervision_.frames_per_sequence, 
+              idx = t % supervision_.frames_per_sequence;
+        int32 pdf_id = xent_post[t][j].first;
+        
+        // Same ordering as for 'chain' models
+        deriv_indexes.push_back(MakePair(idx * supervision_.num_sequences + seq, pdf_id));
+        deriv_data.push_back(xent_post[t][j].second);
+      }
+    }
+    CuArray<Int32Pair> cu_deriv_indexes(deriv_indexes);
+    xent_output_deriv_->AddElements(supervision_.weight, cu_deriv_indexes, deriv_data.data());
+  }
+
+
+  this_stats.tot_den_count += tot_den_post;
+  this_stats.tot_num_count += tot_num_post;
+
+  if (this_stats.AccumulateGradients()) 
```

Comment:
there is no need to have a function to tell you whether to accumulate these gradients.  just check the config value.  it will be easier if you have just one config class rather than two.